Allow for Email field to be parsed when receiving mandrill event

Uelb opened the pull request:
When the recipient of the email is in the bcc field, mandrill does not send the mail
address as event[:bcc] but provides a way to retrieve the email address that
received the event, event[:email]. By using this field, and checking if
the address is not present in the to or cc field of the email, we can
retrieve the bcc email address.
Related to #8, solves #19

wingrunr21 commented:
I've got a few comments:

* This is not future proof. As other Griddler adapters do ship a BCC attribute, if Mandrill starts shipping one in the future then this will stomp on it
* You don't need to search `to` and `cc` fields for the email. The griddler controller passes the params hash directly to `normalize_params`. You have direct access to the actual `email` attribute coming from Mandrill
* It would be better if `normalize_params` passed the `email` attribute directly. This could then be accessed via the Griddler Email object's `params[:email]`.
* Please don't update version numbers for gems you don't maintain. Your CHANGELOG update is unnecessary.

Uelb commented:
Thank you for your comments,
* I reverted the changed in the CHANGELOG, sorry for that.
* I will implement tests for this, and come back to you.
* About the future proof comment, I actually think it is a choice for mandrill not to have this bcc field, as it is not a regular field. From the recipient point of view, there can be only one recipient email address in the bcc field, and you do not actually see the bcc field when you receive an email address.
* About searching the to and cc fields, I'm not sure I understand the comment. I actually check if the email address is not already present in the to and cc field, in which case, it should not be considered as a bcc recipient.
* About the email attribute, I'll add that right now. I saw only this line https://github.com/thoughtbot/griddler/blob/master/lib/griddler/email.rb#L6 so I thought that the params object was not available, my bad. I don't know if using the bcc field is relevant if the email param is used.

wingrunr21 commented:
Can you see the bcc header in the `raw_headers` hash? Mandrill should still pass that part along with the rest of the original message headers. If they do, then Griddler can extract that attribute directly from the raw headers. You are sure the `bcc` attribute is not populated?

`params` is available as a private attribute (https://github.com/thoughtbot/griddler/blob/master/lib/griddler/email.rb#L33). We could also augment `Griddler::Email` with a `mandrill_email` attribute that pulls `params[:email]`.

Setting the bcc in this way is different behavior than what other Griddler adapters have implemented. As bcc is a normalized param while `email` is not, I would much prefer `email` be placed on the `params` attribute of Email rather than overriding the behavior, whether or not it can be surmised that `email` was the bcc attribute.

Uelb commented:
I just checked and the bcc attribute is not populated by mandrill. I was actually inspired by https://github.com/Xenapto/griddler-cloudmailin/blob/8ce0f6b0223e7b986a81a96511d512c479b12171/lib/griddler/cloudmailin/adapter.rb#L45
And by https://github.com/hgcode/griddler-sendgrid/blob/master/lib/griddler/sendgrid/adapter.rb#L38
Who both seems to do it that way (checking if the email address is not in the to nor the cc field)
I think that accessing this email with a public method is quite important. Because right now, we have no way of knowing the email address that the sender used when receiving an email if it is not in the to or the cc field, so adding a method to Griddler::Email might be a good idea if you don't plan on using the bcc one.

wingrunr21 commented:
Alright. If you can rebase these commits to make them a bit cleaner then I can merge this.

Uelb commented:
I just rebased my commits (didn't know the power of that command, just love it).

Uelb commented:
I did a mistake on the condition to resolve the bcc email that I just fixed, is there a way to reopen a pull request, I also added a test checking this behavior.
